### PR TITLE
Fixes #1255: Introduce an activation lock to protect the raw connecti…

### LINKED ISSUE #1255: TSAN race in tcp_lite during system_tests_tcp_adaptor
```
WARNING: ThreadSanitizer: data race (pid=7230)
  Write of size 8 at 0x7b400002c240 by thread T6:
    #0 close_raw_connection_XSIDE_IO /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:409 (skrouterd+0x463872)
    #1 close_raw_connection_XSIDE_IO /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:399 (skrouterd+0x463872)
    #2 close_connection_XSIDE_IO /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:418 (skrouterd+0x47dfd7)
    #3 on_connection_event_CSIDE_IO /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:1346 (skrouterd+0x482949)
    #4 handle_event_with_context /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:813 (skrouterd+0x52aefd)
    #5 do_handle_raw_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:819 (skrouterd+0x52aefd)
    #6 handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1100 (skrouterd+0x52aefd)
    #7 thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1145 (skrouterd+0x52c815)
    #8 _thread_init /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:206 (skrouterd+0x4a8c5d)

  Previous read of size 8 at 0x7b400002c240 by thread T2:
    #0 CORE_activate /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:1415 (skrouterd+0x47df9e)
    #1 qdr_activate_connections_CT /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/router_core_thread.c:100 (skrouterd+0x4f8fe2)
    #2 router_core_thread /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/router_core_thread.c:263 (skrouterd+0x4f8fe2)
    #3 _thread_init /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:206 (skrouterd+0x4a8c5d)

  Location is heap block of size 256 at 0x7b400002c200 allocated by thread T5:
    #0 posix_memalign <null> (libtsan.so.2+0x3f988)
    #1 qd_alloc /home/runner/work/skupper-router/skupper-router/skupper-router/src/alloc_pool.c:344 (skrouterd+0x47110a)
    #2 new_tcplite_connection_t /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:52 (skrouterd+0x483529)
    #3 handle_first_outbound_delivery_CSIDE /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:952 (skrouterd+0x483529)
    #4 CORE_deliver_outbound /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:1523 (skrouterd+0x483529)
    #5 qdr_link_process_deliveries /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/transfer.c:180 (skrouterd+0x504b15)
    #6 CORE_push /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:1511 (skrouterd+0x483383)
    #7 qdr_connection_process /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/connections.c:442 (skrouterd+0x4d5194)
    #8 on_core_activate_TIMER_IO /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/tcp_lite/tcp_lite.c:132 (skrouterd+0x464ce7)
    #9 qd_timer_visit /home/runner/work/skupper-router/skupper-router/skupper-router/src/timer.c:318 (skrouterd+0x522db1)
    #10 handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1016 (skrouterd+0x52ae7c)
    #11 thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1145 (skrouterd+0x52c815)
    #12 _thread_init /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:206 (skrouterd+0x4a8c5d)
```

### PR DESCRIPTION
**Pull request by ganeshmurthy**
…on from being activated when it is being torn down

### CODE REVIEW

File: src/adaptors/tcp_lite/tcp_lite.c [hunk 1]
```diff
@@ -1410,10 +1418,12 @@ static void CORE_activate(void *context, qdr_connection_t *core_conn)
 
     case TL_CONNECTION:
         conn = (tcplite_connection_t*) common;
+        sys_mutex_lock(&conn->activation_lock);
         if (IS_ATOMIC_FLAG_SET(&conn->raw_opened)) {
             SET_ATOMIC_FLAG(&conn->core_activation);
```
Comment (ted-ross) [inline src/adaptors/tcp_lite/tcp_lite.c:1423]: With this lock in place, are there any atomic flags we can convert to bools?

Comment (ganeshmurthy) [reply src/adaptors/tcp_lite/tcp_lite.c:1423]: It looks like the `conn->raw_opened` flag can be knocked down to boolean. The `conn->core_activation` flag needs to remain atomic.

Comment (ganeshmurthy) [reply src/adaptors/tcp_lite/tcp_lite.c:1420]: I have added code to this PR that set the `conn->raw_opened` to boolean. Please take a look.

Comment (ganeshmurthy) [reply src/adaptors/tcp_lite/tcp_lite.c:1423]: After further analysis, the atomic flags that are used inside the new activation lock are also sometimes used outside the activation lock, so my conclusion is that those flags cannot be converted from atomic to boolean

File: src/cutthrough_utils.c [hunk 2]
```diff
@@ -64,9 +64,11 @@ static void activate_connection(qd_message_activation_t *activation, qd_directio
 
     case QD_ACTIVATION_TCP: {
         tcplite_connection_t *conn = safe_deref_tcplite_connection_t(activation->safeptr);
+        sys_mutex_lock(&conn->activation_lock);
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply src/cutthrough_utils.c]: > I think you'll have to move the "!!conn" part of the if() outside the mutex calls.

All set. Please take a look.

**Review by kgiusti — APPROVED** [review]

**Pull request merged**